src/time.rs

```diff
@@ -0,0 +1,73 @@
+use std::time::Duration;
+
+use bevy::prelude::*;
+
+use crate::{
+    AdvanceWorld, AdvanceWorldSet, CloneStrategy, ResourceSnapshotPlugin, RollbackFrameCount,
+    DEFAULT_FPS,
+};
+
+/// [`Resource`] describing the rate at which the [`AdvanceWorld`] will run.
+#[derive(Resource, Clone, Copy, Debug, Hash, Deref)]
+pub struct RollbackFrameRate(pub(crate) usize);
+
+impl Default for RollbackFrameRate {
+    fn default() -> Self {
+        Self(DEFAULT_FPS)
+    }
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+pub struct GGRSTime;
+
+/// This plugins provides [`Time<GGRSTime>`], which is rolled-back automatically, and will also
+/// automatically replace [`Time<()>`] when accessed inside [`GgrsSchedule`](`crate::GgrsSchedule`).
+pub struct GGRSTimePlugin;
+
+impl GGRSTimePlugin {
+    pub fn update(
+        mut time: ResMut<Time<GGRSTime>>,
+        framerate: Res<RollbackFrameRate>,
+        frame: Res<RollbackFrameCount>,
+    ) {
+        let this_frame = frame.0 as u64;
+        let framerate = framerate.0 as u64;
+
+        // 1_000_000_000 fits within a u32, and so does frame, making their product at most u64 in size
+        // By scaling to nanoseconds, rounding error should be minimised.
+        let runtime = Duration::from_nanos(this_frame * 1_000_000_000 / framerate);
+
+        time.advance_to(runtime);
+    }
+
+    pub fn replace_default_with_ggrs(
+        mut default_time: ResMut<Time<()>>,
+        ggrs_time: Res<Time<GGRSTime>>,
+    ) {
+        *default_time = ggrs_time.as_generic();
+    }
+
+    pub fn replace_default_with_virtual(
+        mut default_time: ResMut<Time<()>>,
+        virtual_time: Res<Time<Virtual>>,
```

Comment:
Maybe a bit theoretical, but can we be completely sure Virtual if the right variant to restore? Maybe not worth it to restore the exact time we overwrite, though...

Reply:
My rationale here is that we add these schedules to the `MainSchedule` with the `GgrsPlugin`, so `Virtual` is the correct one in the default case. However, you are correct that if instead we used the `FixedUpdate` schedule this would be wrong.

Reply:
Alright, I'm probably worrying too much :)